# Clasificador de sonrisa con Tensorflow - Keras

Requisitos:
<ul>
<li>Tensorflow 2.0</li>
<li>Numpy</li>
<li>OpenCv</li>
</ul>

El siguiente notebook es un ejemplo de la utilización de una red convolucional con keras utilizando los mismos datos del clasificador de sonrisa.

Los datos utilizados se encuentran en la siguiente ruta: 
<a href="https://github.com/abatista667/RedNeuronalTFM/tree/master/datos">https://github.com/abatista667/RedNeuronalTFM/tree/master/datos</a>

In [1]:
import os
#os.environ["CUDA_VISIBLE_DEVICES"]="-1" 
import tensorflow as tf
import numpy as np
from os import walk
import cv2

Esta función permite cargar un conjunto de imágenes y serán clasificada con el parámetro "y"

In [2]:
def getData(path, y, size =(100,100)):
    X =[]
    Y =[]
    for (dirpath, dirnames, filenames) in walk(path):
        for file in filenames:
            img = cv2.imread(path + file)
            img = cv2.resize(img, size)
            X.append(img)
            Y.append(y)
    X = np.array(X, dtype=float)
    Y = np.array(Y, dtype=float)
    return X, Y

Dada una ruta (path) esta función cargara todas las imágenes dentro de las sub carpetas 0 y 1, y devolverá los arreglos "X","Y" correspondientes.

In [4]:
def getBinaryDataSet(path):
    size = (100,100)
    X =[]
    Y =[]
    data = getData(path+"0\\", [0], size)
    X = data[0]
    Y = data[1]
    data = getData(path+"1\\", [1], size)
    X = np.append(X, data[0], axis=0)
    Y = np.append(Y, data[1], axis=0)
    
    return X, Y

In [3]:
#modificar esta ruta
traininigPath = "C:\\Users\\ariel\\OneDrive\\Documentos\\dataset\\sonrisas\\train\\"

Cargar el dataset de entrenamiento.

In [5]:
xTrain, yTrain = getBinaryDataSet(traininigPath)

In [6]:
xTrain.shape

(400, 100, 100, 3)

In [7]:
yTrain.shape

(400, 1)

In [8]:
#modificar esta ruta
testPath = "C:\\Users\\ariel\\OneDrive\\Documentos\\dataset\\sonrisas\\test\\"

Cargar el data set de prueba.


In [9]:
xTest, yTest = getBinaryDataSet(testPath)

Definición del modelo de red convolucional, entrenamiento y evaluación

In [22]:
redConvolucional = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(64, (3, 3), activation="relu"),
    tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(256, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(1, activation='sigmoid')
])
redConvolucional.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [23]:
redConvolucional.fit(xTrain, yTrain, epochs=20)

Train on 400 samples
Epoch 1/20
400/400 [==============================] - 1s 2ms/sample - loss: 1233.3805 - accuracy: 0.5100
Epoch 2/20
400/400 [==============================] - 1s 1ms/sample - loss: 102.3498 - accuracy: 0.46250s - loss: 121.4510 - accuracy: 0.
Epoch 3/20
400/400 [==============================] - 1s 1ms/sample - loss: 33.2330 - accuracy: 0.5625
Epoch 4/20
400/400 [==============================] - 1s 1ms/sample - loss: 9.9800 - accuracy: 0.7250
Epoch 5/20
400/400 [==============================] - 1s 1ms/sample - loss: 4.6737 - accuracy: 0.7675
Epoch 6/20
400/400 [==============================] - 1s 1ms/sample - loss: 0.7928 - accuracy: 0.9200
Epoch 7/20
400/400 [==============================] - 1s 1ms/sample - loss: 0.2299 - accuracy: 0.9550
Epoch 8/20
400/400 [==============================] - 1s 1ms/sample - loss: 0.1679 - accuracy: 0.9700
Epoch 9/20
400/400 [==============================] - 1s 1ms/sample - loss: 0.0700 - accuracy: 0.9775
Epoch 10/20
400/400 [

In [24]:
redConvolucional.evaluate(xTest,yTest)

82/82 [==============================] - 0s 1ms/sample - loss: 0.6293 - accuracy: 0.8780


[0.6292555070504909, 0.8780488]

<h3>Conclusión</h3>

Luego de varios intentos, y diversas configuraciones, en el mejor de cada uno de los casos, prácticamente no hubo en los resultados obtenidos.

En el perceptrón multicapa el resultado fue de un 87.7% de aciertos mientras que la red convolucional fue de un 87.8% de aciertos.